app/src/items.c

```
@@ -0,0 +1,522 @@
+ *  limitations under the License.
+ ********************************************************************************/
+#include "items.h"
+
+#include <base64.h>
+
+#include "crypto_helper.h"
+#include "items_format.h"
+#include "parser_impl.h"
+
+static items_error_t items_storeSigningTransaction();
+static items_error_t items_storeNetwork();
+static items_error_t items_storeRequiringCapabilities();
+static items_error_t items_storeKey();
+static items_error_t items_validateSigners();
+static items_error_t items_storePayingGas();
+static items_error_t items_storeAllTransfers();
+static items_error_t items_storeCaution();
+static items_error_t items_storeChainId();
+static items_error_t items_storeUsingGas();
+static items_error_t items_checkTxLengths();
+static items_error_t items_storeHash();
+static items_error_t items_storeSignForAddr();
+static items_error_t items_storeGasItem(uint16_t json_token_index);
+static items_error_t items_storeTxItem(parsed_json_t *json_all, uint16_t transfer_token_index, uint8_t *num_of_transfers);
+static items_error_t items_storeTxCrossItem(parsed_json_t *json_all, uint16_t transfer_token_index,
+                                            uint8_t *num_of_transfers);
+static items_error_t items_storeTxRotateItem(parsed_json_t *json_all, uint16_t transfer_token_index);
+static items_error_t items_storeUnknownItem(parsed_json_t *json_all, uint16_t transfer_token_index);
+
+#define MAX_ITEM_LENGTH_TO_DISPLAY 256
+
+item_array_t item_array;
+
+uint8_t hash[BLAKE2B_HASH_SIZE] = {0};
+
+char base64_hash[44];
+
+void items_initItems() {
+    MEMZERO(&item_array, sizeof(item_array_t));
+
+    item_array.numOfUnknownCapabitilies = 1;
+
+    for (uint8_t i = 0; i < sizeof(item_array.items) / sizeof(item_array.items[0]); i++) {
+        item_array.items[i].can_display = bool_true;
+    }
+}
+
+item_array_t *items_getItemArray() { return &item_array; }
+
+void items_storeItems() {
```

items_storeItems should return an error, something like:
```c
items_error_t items_storeItems() {
     CHECK_ERROR(items_storeSigningTransaction());
     CHECK_ERROR(items_storeNetwork());
```
and so on?

---

Wouldn't that make the parsing stop if anything goes wrong? According to the test vectors, when there is an invalid value we should either not display it, show it as unknown capability or raise a warning, depending on the value. What do you think?